Add typless client side GetIndexRequest calls and response class

cbuescher opened the pull request:
The HLRC client currently uses `org.elasticsearch.action.admin.indices.get.GetIndexRequest` 
and `org.elasticsearch.action.admin.indices.get.GetIndexResponse` in its get index calls. Both request and
response are designed for the typed APIs, including some return types e.g. for `getMappings()` which in
the maps it returns still use a level including the type name.
In order to change this without breaking existing users of the HLRC API, this PR introduces two new request
and response objects in the `org.elasticsearch.client.indices` client package. These are used by the 
IndicesClient#get and IndicesClient#exists calls now by default and support the type-less API. The old request
and response objects are still kept for use in similarly named, but deprecated methods.
The newly introduced client side classes are simplified versions of the server side request/response classes since
they don't need to support wire serialization, and only the response needs fromXContent parsing (but no 
xContent-serialization, since this is the responsibility of the server-side class). 
Also changing the return type of `GetIndexResponse#getMapping` to 
`ImmutableOpenMap<String, MappingMetaData> getMappings()`, while it previously was returning another map 
keyed by the type-name.

Commit:
Introduce client GetIndexRequest and Response

Commit:
Trim down copies of Request and Response

Commit:
Adding unit tests for new client side classes

elasticmachine commented:
Pinging @elastic/es-search

elasticmachine commented:
Pinging @elastic/es-core-features

cbuescher commented:
@elasticmachine run elasticsearch-ci/1

cbuescher commented on server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java:
```diff
@@ -61,7 +60,7 @@
     private String[] indices;
 
-    GetIndexResponse(String[] indices,
+    public GetIndexResponse(String[] indices,
```
Needed this access for tests.

cbuescher commented on server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java:
```diff
@@ -315,9 +314,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                indexMappings.put(mappingType, new MappingMetaData(mappingType, parser.map()));
+            } else if (parser.currentToken() == Token.START_ARRAY) {
+                parser.skipChildren();
+            }
```
This changes the parsing of the server-side response back to accept the "old" response format that is including types. I did this because the "new" format is now parsed in the client-side GetIndexResponse, while this parsing code should only be used by the HLRC deprecated client code that uses this GetIndexResponse, and that forces "include_type_name=true" all the time. On the server side we are only using this parsing code in round-trip tests (GetIndexResponseTests)

jtibshirani replied:
Should we also undo the changes to `toXContent`, so that we always include the type name? I no longer think we need to support both the `include_type_name = true` and `include_type_name = false` paths. I think this will make reasoning about these classes quite a bit easier.

cbuescher replied:
I think we do, wasn't this the whole point of leaving the user the option to set this parameter to switch back to the new behaviour in 7.0? Leaving the HLRC aside, any user of the REST API should still be able to ask for both response versions, which will both be rendered by this client side response class.

jtibshirani replied:
I'm sorry, you're completely right! I got confused because I have been staring at the 'put' side of things, and forgot that these classes actually render the REST responses.

cbuescher commented on server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java:
```diff
@@ -203,4 +204,13 @@ public void testCanOutput622Response() throws IOException {
+    @Override
+    protected ToXContent.Params getToXContentParams() {
+        return new ToXContent.MapParams(Collections.singletonMap(BaseRestHandler.INCLUDE_TYPE_NAME_PARAMETER, "true"));
+    }
```
This is forced to "true" now because that is what all the calls from the HLRC using this server side Request/Repsone pair are doing.

jtibshirani replied:
Related to my question above, can we undo the changes in the server-side `GetIndexResponse#toXContent` so that we always include the type name?

cbuescher replied:
See my thinking above.

jtibshirani commented on server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java:
```diff
@@ -112,8 +114,7 @@ private GetIndexResponse createTestInstance(boolean randomTypeName) {
 
     @Override
     protected GetIndexResponse createXContextTestInstance(XContentType xContentType) {
```
Can we simplify this test now to have a single `createTestInstance` method, and avoid this extra override?

cbuescher commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java:
```diff
@@ -173,7 +172,22 @@ public void testIndicesExists() throws IOException {
+
+    public void testIndicesExistsWithTypes() throws IOException {
+        // Index present
+        {
```
nope, will change.

cbuescher replied:
Actually, its what the formater does by default. We do this in other tests in this class as well it seems to keep local variable of the individual cases separate. I'd prefer leaving it although I would do this differently in another new test.

Commit:
Correct indention

cbuescher commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesRequestConvertersTests.java:
```diff
@@ -123,7 +123,35 @@ public void testIndicesExistEmptyIndices() {
+
+    public void testDeprecatedIndicesExistEmptyIndices() {
+        LuceneTestCase.expectThrows(IllegalArgumentException.class, ()
+            -> IndicesRequestConverters.indicesExist(new GetIndexRequest()));
```
I think this is okay, but I added tests for the deprecated converter method as well.

Commit:
Add tests for deprecated methods

Commit:
Simplify o.e.action.a..GetIndexResponseTests

Commit:
Merge branch 'master' into client-side-getIndexRequest

cbuescher commented:
@jtibshirani thanks for the reviews, I pushed a round of changes hopefully adressing your comments.

jtibshirani commented on client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetIndexResponseTests.java:
```diff
@@ -0,0 +1,176 @@
+     *
+     * TODO: ideally we would create a random instance of the client side response that contains the rendering code
+     * write it to xContent, parse it back and compare the results
+     * this currently would mean we would need test dependencies to server:test
```
I'm not sure I really understand this TODO. For what purpose would we need a dependency on `server:test`?

jtibshirani replied:
An update: as I was working on the same refactor for 'get mappings', I found it nice to use an `xContentTester`. I ended up creating a method `GetMappingsResponseTests#toXContent` that converts an HLRC `GetMappingsResponse` to a server-side one, then calls `toXContent` on that. Here's the test class in case you'd like to take a look: https://github.com/elastic/elasticsearch/pull/37796/files#diff-f66406b858061334c3239c8d2148b2caR41

cbuescher replied:
Thanks, I'll take a look...

cbuescher replied:
I see what you did there, I think its basically the same as my approach, although I have to say I really dislike using  xContentTester for all the fuss around having to set up functions for everything. If you've used it several times it becomes clearer what it does, but its over-engineered from my point of view. I can change to using it if you really like though.
The main point though is to get an xContent rendering of the server side response. While you are generating a random instance of the client-side class, then convert it to server-side to call its xContent, I copied the code to create a randomized server-side response directly from `org.elasticsearch.action.admin.indices.get.GetIndexResponseTests`. I think it boild down to the same disadvantage, namely having to copy randomization code (which you also seem to have copied over, at least partially). By my comment I mean it would be great if we could call `org.elasticsearch.action.admin.indices.get.GetIndexResponseTests#createTestInstance` in the client-side test to be sure we get a "real" server side response with all its randomization variants, and call its "real" toXContent, because that is what the client is going to have to parse. In order to avoid code changing on the server side unnoticed it would be good to actually use that code, not rely on the copy being up to date which we have to do if we copy. Does this make sense?

The problem I'm running into here is that the client project can see `org.elasticsearch.action.admin.indices.get.GetIndexResponse` but not `org.elasticsearch.action.admin.indices.get.GetIndexResponseTests.` Maybe we could have a test dependency for this? But @hub-cap might know. We can probably do this in a follow up though if people think its useful to have.

jtibshirani replied:
Thanks, I understand your comment now! Doing a bit more digging, I see `AbstractHlrcXContentTestCase`, which I think was designed for this situation. This sort of class tests both server serialization and HLRC serialization at the same time, which means there is a single place to generate random server objects. HLRC serialization is tested by generating a random server object, calling `toXContent` on it, and parsing it back as an client object. However, this class currently lives in the x-pack directory, which doesn't seem like a natural place to move `GetMappingsResponseTests`, `GetIndexResponseTests`, etc.

A proposal for what to do with our two open PRs:
- We can plan to follow-up with @hub-cap to discuss where to move `AbstractHlrcXContentTestCase`, then refactor `Get*Tests` to all extend this class.
- For now we can generate random server objects (as you do in this PR). I'll switch my PR to do that instead of generating a client-side one, so we are closer to modeling what actually happens during an HLRC call.
- I would prefer us to use `xContentTester`, because although it's a bit overkill, it is used extensively and I think being consistent with the test framework leads to overall less confusion than being inconsistent but slightly simpler in one or two tests.

jtibshirani commented on client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetIndexResponseTests.java:
```diff
@@ -0,0 +1,176 @@
+        );
+    }
+
+    public static ImmutableOpenMap<String, MappingMetaData> createMappingsForIndex(int typeCount) {
```
To avoid duplicating this logic, would it be possible to pull out a test utilities class similar to `RandomCreateIndexGenerator` that lives in `test:framework`? This would also be useful to have available when working on 'get mappings'.

jtibshirani reviewed (commented):
Sorry for the confusion in my previous review, this is now looking good to me from the types deprecation side. I just left a couple questions around tests.

jtibshirani commented on client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesRequestConvertersTests.java:
```diff
@@ -123,7 +123,40 @@ public void testIndicesExistEmptyIndices() {
+    }
+
+    public void testDeprecatedIndicesExistEmptyIndices() {
+        LuceneTestCase.expectThrows(IllegalArgumentException.class, ()
```
Sorry if I'm missing something, but I think these first two lines are just a repeat of ` testIndicesExistEmptyIndices`? Also, would it be correct to call this `testIndicesExistEmptyIndicesWithTypes`?

cbuescher replied:
No, you are right. I forgot about the other method, moved them closer to each other now and changed the name.

Commit:
Better naming of testDeprecatedIndicesExistEmptyIndices

cbuescher commented:
@jtibshirani thanks, I adressed one of your comments, left a response to the other. Let me know if we should discuss this again later.

Commit:
Merge branch 'master' into client-side-getIndexRequest

Commit:
Fix compile issue

Commit:
Fix checkstyle

jtibshirani reviewed (approved):
I left one comment about refactoring the test to use `xContent`, to be consistent with our overall testing approach. However this is looking very nice and I'm approving now so as not to hold up merging after the test changes.

Commit:
Merge branch 'master' into client-side-getIndexRequest

cbuescher commented:
> not to hold up merging after the test changes

@jtibshirani  sorry, I'm not sure if this refers to changes you want me to still make or changes we can do as a follow re our discussion in https://github.com/elastic/elasticsearch/pull/37778#discussion_r. Reading that comment I don't see any changes needed for this PR, but will wait for confirmation just before merging.

Commit:
Merge branch 'master' into client-side-getIndexRequest

Commit:
Merge branch 'master' into client-side-getIndexRequest

Commit:
Adapting test

Commit:
Merge branch 'master' into client-side-getIndexRequest

Commit:
Merge branch 'master' into client-side-getIndexRequest

cbuescher commented:
@jtibshirani I updated the test in 6a8a8a9. I have to admit I'm not very happy with the way we test this, but I tried following your suggestions from https://github.com/elastic/elasticsearch/pull/37778#discussion_r251261828 to get this PR merged. I think we should iterate on the testing later on.
Furthermore I simplified most getters for the new client-side GetIndexResponse to return plain Maps instead of the ImmutableOpenMaps that the server side response returns. I saw e.g the new client side GetMappingsResponse does the same so I adapted this approach. Let me know if this works for you.

Commit:
iter

jtibshirani commented:
Thank you @cbuescher for making these updates, they look good to me. Sounds good about iterating on the approach to testing -- at least the strategy is consistent now across the different HLRC classes, which will make it easier to do a follow-up refactor.
